### DISCUSSION TITLE
Is it possible to combine Zeroshot and Semi-supervised approaches?

### DISCUSSION THREAD

**Discussion by ITAYC0HEN** [MaartenGr/BERTopic#1895] [Q&A]
When trying to execute `fit_transform(docs, y=categories)` with Zeroshot and with Semi-supervised combined, I am receiving an error that says the length of X and the length of Y should be the same.

Naturally, when passed docs and categories, I verified that `len(docs)==len(categories)`.

By reading the code, it seems like Zeroshot is being executed prior to the semi-supervised code, and somehow reduces the size of either x or y 

Is there an option to execute BERTopic with both zeroshot and semi-suprvised? I want both functionalities to be included in the topic creation process.
If so, how?

My current code looks something like this:

```python
        docs = data["data"]
        categories = data["target"]

        zero_shot_topics = ["My first topic", "My second topic" , ....]

        topic_model = BERTopic(
            < ... >
            zeroshot_topic_list=zero_shot_topics,
            zeroshot_min_similarity=.6,

        )

    topics, probs = topic_model.fit_transform(docs, y=categories)
```

Thank you!

----

**Exception frames:**
```

[ ... ]


[/usr/local/lib/python3.10/dist-packages/bertopic/_bertopic.py](https://localhost:8080/#) in fit(self, documents, embeddings, images, y)
    314         ```
    315         """
--> 316         self.fit_transform(documents=documents, embeddings=embeddings, y=y, images=images)
    317         return self
    318 

[/usr/local/lib/python3.10/dist-packages/bertopic/_bertopic.py](https://localhost:8080/#) in fit_transform(self, documents, embeddings, images, y)
    406 
    407         # Reduce dimensionality
--> 408         umap_embeddings = self._reduce_dimensionality(embeddings, y)
    409 
    410         # Cluster reduced embeddings

[/usr/local/lib/python3.10/dist-packages/bertopic/_bertopic.py](https://localhost:8080/#) in _reduce_dimensionality(self, embeddings, y, partial_fit)
   3353                 # cuml umap needs y to be an numpy array
   3354                 y = np.array(y) if y is not None else None
-> 3355                 self.umap_model.fit(embeddings, y=y)
   3356             except TypeError:
   3357 

[/usr/local/lib/python3.10/dist-packages/umap/umap_.py](https://localhost:8080/#) in fit(self, X, y, force_all_finite)
   2666             len_X = len(X) if not self._sparse_data else X.shape[0]
   2667             if len_X != len(y):
-> 2668                 raise ValueError(
   2669                     "Length of x = {len_x}, length of y = {len_y}, while it must be equal.".format(
   2670                         len_x=len_X, len_y=len(y)

ValueError: Length of x = 11752, length of y = 11779, while it must be equal.
```

**Comment by MaartenGr** [comment]
Based on the following error:

> ValueError: Length of x = 11752, length of y = 11779, while it must be equal.

Did you make sure that `len(categories) == len(docs)` because the error suggests that it is not.

**Reply by ITAYC0HEN** [reply]
Hey Maarten, thanks for the quick reply! Absolutely, as mentioned in the post

> Naturally, when passed docs and categories, I verified that len(docs)==len(categories).

So yeah, I checked and verified :)

I also checked the source code and it seems that zero-shot might modify the value of `y`

https://github.com/MaartenGr/BERTopic/blob/1002deee5a0402ed7e69ceed0dded064697a59fb/bertopic/_bertopic.py#L401-L411

**Reply by MaartenGr** [reply]
Where do you see the value of `y` to be modified? It is used in `_reduce_dimensionality` and `_cluster_embeddings` but I do not remember `y` being updated.

**Reply by ITAYC0HEN** [reply]
Sorry, you are correct. It seems like `x` is changed during the execution of either , `x` being the embeddings. This happens during the execution of `_zeroshot_topic_modeling`.

I think that `x` becomes the number of unassigned embeddings, meaning, docs that were not able to assign to zero-shot topics.
It splits the embeddings to two arguments, `embeddings` and `assigned_embeddings`.
https://github.com/MaartenGr/BERTopic/blob/1002deee5a0402ed7e69ceed0dded064697a59fb/bertopic/_bertopic.py#L403

But then, to `_reduce_dimensionality` it only sends  `embeddings`. This makes `len(embeddings)` to be lower than `len(y)`.
https://github.com/MaartenGr/BERTopic/blob/1002deee5a0402ed7e69ceed0dded064697a59fb/bertopic/_bertopic.py#L408

**Reply by MaartenGr** [reply]
You're right! It seems that I split up the embeddings for the two use cases which cannot then be used in `_reduce_dimensionality` if `y` is also used. 

It might be best here to also split up `y` to match the assignment of `embeddings` as is done here:

https://github.com/MaartenGr/BERTopic/blob/424cefc68ede08ff9f1c7e56ee6103c16c1429c6/bertopic/_bertopic.py#L3572

At the moment, I have limited time to work on this but if you want, a PR would be much appreciated!
